### ISSUE TITLE
Built-in pdf viewer damages some urls

### ISSUE THREAD

**Issue by i3v** [texstudio-org/texstudio#979]
### Environment
- TeXstudio:  2.12.22 (git 2.12.22)
- Qt: Qt Version 5.12.1, compiled with Qt 5.12.1 R
- OS: Windows 10
- TeX distribution: TexLive 2019

Also tested in "TeXstudio 2.12.16 (git 2.12.16)" - the issue is the same.

### Expected behavior
Both links in this pdf ([viewer_damages_url_example.pdf](https://github.com/texstudio-org/texstudio/files/4320323/viewer_damages_url_example.pdf)) should work OK.

### Actual behavior
Although pdf itself is OK (e.g. both links work if I click on them in Foxit Reader or Adobe Reader), TexStudio built-in pdf viewer does not open the correct page (in my web browser) when I click on the first link.

### How to reproduce

 

1.  Build the following tex code (or just download the attached pdf):
```
\documentclass{article}
\usepackage{hyperref}
\begin{document}
\hyperref{https://www.scopus.com/inward/record.uri?eid=2-s2.0-85063442214\&doi=10.1117\%2f12.2522991\&partnerID=40\&md5=945b4b60e9b05578fd98c79bfba20c9a}{cat}{name}{this link does not work in TexStudio pdf viewer}

\hyperref{https://www.scopus.com/inward/record.uri?eid=2-s2.0-85079950288\&partnerID=40\&md5=e5b3d225e63fe4df1dfd4ebbfc3dbce6}{cat}{name}{this link is somewhat similar but works OK}
\end{document}
```
2. Just for reference, open pdf in Adobe Acrobat / Foxit Reader / whatever, click on both links.
3. Open pdf in TexStudio built-in pdf viewer, click on both links.

### The issue
It looks like TexStudio pdf viewer somehow damages the first url while passing it to the web browser: `%2f` is replaced with `%2F`, thus, we end up on a kind of error page (it asks you to log in), instead of the actual target (just shows you the info about the paper).

Both links once again, just for reference:
1. https://www.scopus.com/inward/record.uri?eid=2-s2.0-85063442214&doi=10.1117%2f12.2522991&partnerID=40&md5=945b4b60e9b05578fd98c79bfba20c9a
2. https://www.scopus.com/inward/record.uri?eid=2-s2.0-85079950288&partnerID=40&md5=e5b3d225e63fe4df1dfd4ebbfc3dbce6

**Comment by muzimuzhi** [comment]
Try this:
```tex
\documentclass{article}
\usepackage{hyperref}
\begin{document}
\href{https://www.scopus.com/inward/record.uri?eid=2-s2.0-85063442214&doi=10.1117\%2f12.2522991&partnerID=40&md5=945b4b60e9b05578fd98c79bfba20c9a}{this link does not work in TexStudio pdf viewer}

\href{https://www.scopus.com/inward/record.uri?eid=2-s2.0-85079950288&partnerID=40&md5=e5b3d225e63fe4df1dfd4ebbfc3dbce6}{this link is somewhat similar but works OK}
\end{document}
```

Notes:
 - Use `\href{<url>}{<text>}` instead of `\hyperref`
 - In your case, `%` is the only character that need to be escaped.

**Comment by i3v** [comment]
OK, just tried that. 
Works just the same as my version.
Although I do admit that your version looks nicer :)

**Comment by muzimuzhi** [comment]
This seems more like a latex problem, not an editor problem, and TeXstudio is an editor. Maybe https://tex.stackexchange.com/ is the right place to ask.

BTW, does upgrading all texlive packages help?

**Comment by sunderme** [comment]
oddly enough, this is a Qt issue.
The problem is the part "...%2f12..." .
Qt automatically changes it to "...%2*F*12..." which is not recognized by the website.
I have not found a way, to change that behavior. Whether it is a bug of Qt or the original url is non-conformant, I can't say. However I don't see a way to fix this.

**Comment by muzimuzhi** [comment]
It seems this is a website issue. 
 - `%2f` is a percent-encoding byte to represent character `/`. 
 - In the living standards of both URI and URL, upper- and lowercase letters used in percent-encoding bytes are equivalent. And the encoding process should consistently use uppercase, so what Qt does is right.

**References**:
   > The uppercase hexadecimal digits 'A' through 'F' are equivalent to
   the lowercase digits 'a' through 'f', respectively.  If two URIs
   differ only in the case of hexadecimal digits used in percent-encoded
   octets, they are equivalent.  For consistency, URI producers and
   normalizers should use uppercase hexadecimal digits for all percent-
   encodings.
   > 
   > RFC3986 URI: Generic Syntax, Sec. 2.1. Percent-Encoding ([link](https://tools.ietf.org/html/rfc3986#section-2.1))

   > To percent encode a byte into a percent-encoded byte, return a string consisting of U+0025 (%), followed by two ASCII upper hex digits representing byte.
   > To percent decode a byte sequence input, ... if byte is 0x25 (%) and the next two bytes after byte in input are not in the ranges 0x30 (0) to 0x39 (9), 0x41 (A) to 0x46 (F), and 0x61 (a) to 0x66 (f), all inclusive, append byte to output. ...
   > 
   > The URL Standard, Sec. 1.3. Percent-encoded bytes ([link](https://url.spec.whatwg.org/#percent-encoded-byte))

**Comment by sunderme** [comment]
okay, so we can‘t fix it

**Comment by i3v** [comment]
>oddly enough, this is a Qt issue.

Indeed, it turns out to be so... This happens even for a simple `QString`->`QUrl`. There are two options:

- `QDesktopServices::openUrl(QUrl(str))`, which is the same as `QDesktopServices::openUrl(QUrl(str,QUrl::TolerantMode))`  - according to the docs, this mode autofixes some common "user-input" issues - for example, it replaces spaces with `%20`. Thus one should expect replacements like `%2f` -> `%2F` too. I was assuming that TexStudio, effectively, uses this mode. (The real TexStudio code is [trickier](https://github.com/texstudio-org/texstudio/blob/37061c365ced545860a9ec6efa6b5430f1abf7f1/src/pdfviewer/PDFDocument.cpp#L1127).)
- `QDesktopServices::openUrl(QUrl(str, QUrl::StrictMode))` - according to it's name, I've been expecting that it should not change url anyhow. But... It still replaces `%2f` with `%2F`! And this is not documented. And there's no "even more strict" mode.

> okay, so we can‘t fix it

Yep... Not with Qt.

I've reported the issue to both Scopus and Qt. I won't expected that any of them would fix their issues anytime soon though.


> It seems this is a website issue.

Indeed, there's a bug on Scopus side. But there's also a bug on ~~TexStudio~~Qt side. Only a combination of both results in a real problem. See [QTBUG-84020](https://bugreports.qt.io/browse/QTBUG-84020), TLDR:

The URL, produced by Scopus does not fully comply with the standard (which is a poor but somewhat common practice). Still, everything works OK (e.g. in Foxit Reader, Acrobat Reader, Microsoft Word, ...,  in _all_ apps I've tested)... until you introduce `QUrl`. Moreover, even Okular and Konqueror (they are from KDE, e.g. related to Qt) do work OK. 

E.g. the common practice is to be lenient to that non-conformance. This is also supported by the following paragraph from RFC3986:
>Under normal circumstances, the only time when octets within a URI are percent-encoded is during the process of producing the URI from its component parts

Be lenient in what you accept, be strict in what you generate. And, since we're not talking about generating an url, no percent-encoding parsing should ever happen. And thus, no case normalization for percent-encoding should happen. All other apps/frameworks do follow this principle. Qt violates it. 
(Sadly, Qt dev disagrees with this.)

Sorry for bumping. Hope these references would be useful for someone sometime.
